<a href="https://colab.research.google.com/github/yasirabd/solver-society-job-data/blob/main/4_Data_cleaning_Glints.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Objectives
Cleaning data glints

Data input yang dibutuhkan:
1. glints_4_oct.csv
2. job_glints_mapper.csv
3. daftar_kota_kabupaten_id.csv
4. glints_company_loc_prov_pulau.csv

Data output yang dihasilkan:
1. data_loker_glints_16oct.csv
2. data_pendidikan_glints_16oct.csv
3. data_skill_glints_16oct.csv


In [1]:
import pandas as pd
import numpy as np
import re

In [2]:
glints = pd.read_csv("/content/drive/My Drive/Data Loker/glints_4oct.csv")
glints.head()

platform  ...                                     skills_to_have
0   glints  ...  Must Have Skills [SEP] .NET [SEP] English [SEP...
1   glints  ...  Must Have Skills [SEP] ReactJS [SEP] HTML [SEP...
2   glints  ...  Must Have Skills [SEP] Teamwork [SEP] Microsof...
3   glints  ...  Must Have Skills [SEP] Agile Scrum [SEP] Engli...
4   glints  ...  Must Have Skills [SEP] Cms Wordpress [SEP] Wor...

[5 rows x 9 columns]

# Data Cleaning

In [3]:
# check null value
glints.isnull().sum()

platform            0
job_position        0
company_name        0
job_category        0
company_location    0
job_salary          0
job_type            0
job_desc            0
skills_to_have      0
dtype: int64

## job_position

In [4]:
job_position = glints['job_position'].str.lower()
job_position.value_counts()

backend developer            25
fullstack developer          15
account executive            13
frontend developer           13
sales executive              12
                             ..
senior fullstack              1
general manager               1
it field support baturaja     1
treasury analyst              1
community engagement          1
Name: job_position, Length: 1433, dtype: int64

Dapat kita bersihkan lagi supaya lebih distinct. Misalkan
value **'backend developer', 'back end developer', 'back-end developer'** dapat kita perbaiki menjadi **'backend developer'**.

### Clustering pekerjaan

In [5]:
def clean_job(keyword, replace_keyword, data):
    for i in range(len(data)):
        if keyword in data[i]:
            data[i] = replace_keyword
    return data

In [6]:
temp = job_position.tolist()

for i, job in enumerate(temp):
    job = re.sub("[\(\[].*?[\)\]]", "", job)
    temp[i] = job

data = clean_job('backend', 'backend developer', temp)
data = clean_job('back end', 'backend developer', data)
data = clean_job('back-end', 'backend developer', data)
data = clean_job('fullstack', 'fullstack developer', data)
data = clean_job('full stack', 'fullstack developer', data)
data = clean_job('full-stack', 'fullstack developer', data)
data = clean_job('frontend', 'frontend developer', data)
data = clean_job('front end', 'frontend developer', data)
data = clean_job('front-end', 'frontend developer', data)
data = clean_job('account executive', 'account executive', data)
data = clean_job('sales executive', 'sales executive', data)
data = clean_job('product manager', 'product manager', data)
data = clean_job('account manager', 'account manager', data)
data = clean_job('android', 'mobile developer', data)
data = clean_job('ios', 'mobile developer', data)
data = clean_job('software engineer', 'software engineer', data)
data = clean_job('ui/ux', 'ui/ux designer', data)
data = clean_job('ux designer', 'ui/ux designer', data)
data = clean_job('business development', 'business development', data)
data = clean_job('project manager', 'project manager', data)
data = clean_job('graphic designer', 'graphic designer', data)
data = clean_job('web developer', 'web developer', data)
data = clean_job('web programmer', 'web developer', data)
data = clean_job('quality assurance', 'quality assurance', data)
data = clean_job('qa', 'quality assurance', data)
data = clean_job('java developer', 'java developer', data)
data = clean_job('digital marketing', 'digital marketing', data)
data = clean_job('data scientist', 'data scientist', data)
data = clean_job('sales manager', 'sales manager', data)
data = clean_job('mobile developer', 'mobile developer', data)
data = clean_job('mobile', 'mobile developer', data)
data = clean_job('mobile app developer', 'mobile developer', data)
data = clean_job('devops engineer', 'devops engineer', data)
data = clean_job('devops', 'devops engineer', data)
data = clean_job('social media', 'social media strategist', data)
data = clean_job('content creator', 'content creator', data)
data = clean_job('accounting staff', 'accounting staff', data)
data = clean_job('graphic design', 'graphic designer', data)
data = clean_job('head of sales', 'head of sales', data)
data = clean_job('purchasing staff', 'purchasing staff', data)
data = clean_job('administrasi', 'staff administrasi', data)
data = clean_job('administration staff', 'staff administrasi', data)
data = clean_job('staff admin', 'staff administrasi', data)
data = clean_job('business consultant', 'business consultant', data)
data = clean_job('software developer', 'software engineer', data)
data = clean_job('marketing executive', 'marketing executive', data)
data = clean_job('sales agent', 'sales', data)
data = clean_job('sales person', 'sales', data)
data = clean_job('salesman', 'sales', data)
data = clean_job('sales development', 'sales', data)
data = clean_job('programmer', 'programmer', data)
data = clean_job('marketing manager', 'marketing manager', data)
data = clean_job('fashion designer', 'fashion designer', data)
data = clean_job('translator', 'translator', data)
data = clean_job('customer service', 'customer service', data)
data = clean_job('telemarketing', 'telemarketing', data)
data = clean_job('it security', 'it security', data)
data = clean_job('seo', 'seo specialist', data)
data = clean_job('hr', 'hrd', data)
data = clean_job('scrum', 'scrum master', data)
data = clean_job('data engineer', 'data engineer', data)

In [7]:
pd.Series(data).value_counts()

fullstack developer                84
mobile developer                   79
backend developer                  79
frontend developer                 60
graphic designer                   50
                                   ..
finance & accounting internship     1
procurement manager                 1
associate consultant                1
accounting finance                  1
team opening                        1
Length: 887, dtype: int64

Sudah cukup distinct, dari **1433** unique data menjadi **887**.

Kita lakukan manual cleaning pada google sheet.

In [8]:
# pd.Series(data).value_counts().to_csv('job_glints2.csv')

j = pd.read_csv("/content/drive/My Drive/Data Loker/job_glints_mapper.csv")

j['target'] = j['target'].str.lower()
j['target'].value_counts()

sales                      36
tutor                      22
marketing                  18
content writer             13
consultant                 12
                           ..
data quality test lead      1
web testing                 1
partnership associate       1
chief financial officer     1
team opening                1
Name: target, Length: 597, dtype: int64

In [9]:
# create dict mapper
job_mapper = dict()

for idx, row in j.iterrows():
    job_mapper[row['asli']] = row['target']

temp = pd.Series(data).map(job_mapper)

In [10]:
temp.isnull().sum()

201

In [11]:
# isi data kosong dengan 'dihapus'
data = temp.fillna('dihapus')

for idx,d in enumerate(data.tolist()):
    if len(d)<=1:
        print(idx,d)

In [12]:
# capitalize
for idx,d in enumerate(data):
    data[idx] = d.title()

pd.Series(data).value_counts()

Dihapus                      201
Fullstack Developer           84
Mobile Developer              79
Backend Developer             79
Sales                         61
                            ... 
Freelance Designer             1
Finance & Tax Officer          1
Creative Group Head            1
Manager Rel Kereta Api         1
Network Operations Center      1
Length: 506, dtype: int64

In [13]:
# pd.Series(temp).value_counts().to_csv('glints_job.csv')
# job_position.to_csv('glints_15oct.csv', index=False)

In [14]:
glints['job_position'] = data

# drop row yang 'Dihapus'
glints = glints[glints.job_position != 'Dihapus'].reset_index(drop=True)
glints.shape

(1749, 9)

## company_name

In [15]:
company_name = glints['company_name']
company_name.value_counts()

PT. Bintang Jaya Unggul Sentosa           23
Confidential Company Talenthunt 1 -006    20
AIA Singapore Pte Ltd                     19
PT Hokinda Citralestari                   19
Jakmall.com                               17
                                          ..
Id/x Partners                              1
Sojilabs Indonesia                         1
Ashnik Pte. Ltd.                           1
PT. Venteny Matahari Indonesia             1
Anacle Systems Limited                     1
Name: company_name, Length: 718, dtype: int64

> tidak diubah

## job_category

In [16]:
job_category = glints['job_category']
job_category.value_counts()

Software Engineering            568
Business Development / Sales    269
Marketing                       186
Design                          125
Finance                          86
Other                            68
Operations                       65
Product Management               63
Administrative                   62
Engineering                      50
Media & Communications           50
Education                        49
Data Science                     35
Human Resource                   30
Consulting                       13
Management                       13
Legal                            11
Hospitality and Travel            3
Healthcare                        2
Language and Translation          1
Name: job_category, dtype: int64

> Sudah cukup bagus

## company_location

In [17]:
company_location = glints['company_location']
company_location.value_counts()

Jakarta, Indonesia            1088
Batam, Indonesia               131
Bandung, Indonesia             113
Yogyakarta, Indonesia           82
Surabaya, Indonesia             64
Tangerang, Indonesia            58
South Tangerang, Indonesia      41
Malang, Indonesia               24
Medan, Indonesia                23
Bekasi, Indonesia               22
Depok, Indonesia                18
Bogor, Indonesia                16
Semarang, Indonesia             14
Tulungagung, Indonesia           6
Padang, Indonesia                6
Denpasar, Indonesia              5
Pekanbaru, Indonesia             4
Serpong, Indonesia               4
Makassar, Indonesia              3
Pandeglang, Indonesia            2
Badung, Indonesia                2
Palembang, Indonesia             2
Solo, Indonesia                  2
Mojokerto, Indonesia             2
Cirebon, Indonesia               2
Balikpapan, Indonesia            2
Cilegon, Indonesia               1
Klaten, Indonesia                1
Abepura, Indonesia  

In [18]:
# company_location.value_counts().to_csv('glints_company_loc.csv')

Kita ekstrak menjadi atribut <code>pulau</code>, <code>provinsi</code>, <code>kota</code>.

In [19]:
# data bantuan 
kota_kab = pd.read_csv("/content/drive/My Drive/Data Loker/daftar_kota_kabupaten_id.csv")
kota_kab.head()

Pulau Provinsi   Kabupaten/Kota
0  Sumatera     Aceh       Aceh Barat
1  Sumatera     Aceh  Aceh Barat Daya
2  Sumatera     Aceh       Aceh Besar
3  Sumatera     Aceh        Aceh Jaya
4  Sumatera     Aceh     Aceh Selatan

Dilakukan manual perbaikan data pada google sheet.

In [20]:
loc = pd.read_csv("/content/drive/My Drive/Data Loker/glints_company_loc_prov_pulau.csv")
loc.head()

Asli     Pulau        Provinsi  company_location
0     Jakarta, Indonesia      Jawa     DKI Jakarta              1088
1       Batam, Indonesia  Sumatera  Kepulauan Riau               131
2     Bandung, Indonesia      Jawa      Jawa Barat               113
3  Yogyakarta, Indonesia      Jawa   DI Yogyakarta                82
4    Surabaya, Indonesia      Jawa      Jawa Timur                64

In [21]:
# create dict mapper
pulau_mapper = dict()
prov_mapper = dict()
for idx, row in loc.iterrows():
    pulau_mapper[row['Asli']] = row['Pulau']
    prov_mapper[row['Asli']] = row['Provinsi']

pulau = company_location.map(pulau_mapper)
provinsi = company_location.map(prov_mapper)

In [22]:
kota = list()
for loc in company_location:
    split = loc.split(',')[0].strip()
    kota.append(split)

glints['kota'] = kota
glints['pulau'] = pulau
glints['provinsi'] = provinsi

# drop company_location
glints = glints.drop(['company_location'], axis=1)

In [23]:
glints.head()

platform         job_position  ...     pulau        provinsi
0   glints       .Net Developer  ...  Sumatera  Kepulauan Riau
1   glints  Fullstack Developer  ...  Sumatera  Kepulauan Riau
2   glints   Frontend Developer  ...      Jawa     DKI Jakarta
3   glints    Backend Developer  ...      Jawa     DKI Jakarta
4   glints   Frontend Developer  ...  Sumatera  Kepulauan Riau

[5 rows x 11 columns]

## job_salary

In [24]:
job_salary = glints['job_salary']
job_salary.value_counts()

Perusahaan tidak menampilkan gaji    1217
IDR 4,000,000 -6,000,000 / month       22
IDR 4,000,000 -5,000,000 / month       16
IDR 3,500,000 -5,000,000 / month       15
IDR 5,000,000 -10,000,000 / month      14
                                     ... 
IDR 800,000 -1,500,000 / month          1
IDR 2,000,000 -15,000,000 / year        1
IDR 4,300,000 -6,500,000 / month        1
IDR 500,000 -750,000 / project          1
IDR 300,000 -440,000 / month            1
Name: job_salary, Length: 229, dtype: int64

In [25]:
# remove space sebelum '-'
result = list()
for sal in job_salary:
    split = [data.strip() for data in sal.split('-')]
    join = '-'.join(split)
    result.append(join)

In [26]:
glints['job_salary'] = result

In [27]:
glints.job_salary[1500]

'IDR\xa03,000,000-5,000,000\xa0/\xa0month'

In [28]:
# remove unwanted things
import re

result = list()
for sal in glints['job_salary']:
    clean = re.sub(r'[^\x00-\x7F]+',' ',sal).encode('utf-8').decode('utf-8','ignore').strip()
    result.append(clean)

In [29]:
glints['job_salary'] = result

In [30]:
glints.job_salary[1500]

'IDR 3,000,000-5,000,000 / month'

In [31]:
glints.head()

platform         job_position  ...     pulau        provinsi
0   glints       .Net Developer  ...  Sumatera  Kepulauan Riau
1   glints  Fullstack Developer  ...  Sumatera  Kepulauan Riau
2   glints   Frontend Developer  ...      Jawa     DKI Jakarta
3   glints    Backend Developer  ...      Jawa     DKI Jakarta
4   glints   Frontend Developer  ...  Sumatera  Kepulauan Riau

[5 rows x 11 columns]

## job_type

In [32]:
job_type = glints['job_type']
job_type.value_counts()

FULL_TIME        1580
INTERNSHIP        105
PROJECT_BASED      40
PART_TIME          24
Name: job_type, dtype: int64

> Data oke

## job_desc
Lakukan ekstrak untuk atribut <code>pendidikan</code>

In [33]:
job_description = glints['job_desc'].str.lower()
job_description

0       requirements: [sep] min diploma / degree in in...
1       what is this job about? [sep] mossery is an e-...
2       front-end spec [sep] have experience with at l...
3       job responsibilities: [sep] • designing and de...
4       [sep] wordpress developer [sep] implement prot...
                              ...                        
1744    go mobile indonesia is a rapidly developing fu...
1745    go mobile indonesia [sep] is a rapidly develop...
1746    get involved in a team of web applications, we...
1747    go mobile indonesia [sep] is a rapidly develop...
1748    we are an edu-tech company delivering better l...
Name: job_desc, Length: 1749, dtype: object

In [34]:
counter = 0
for desc in job_description:
    if 'diploma' in desc:
        counter +=1
    elif 'bachelor' in desc:
        counter+=1
    elif 's1' in desc:
        counter+=1
    elif 'd3' in desc:
        counter+=1
    elif 'master degree' in desc:
        counter+=1
    elif 'masters degree' in desc:
        counter+=1
    elif 'degree' in desc:
        counter+=1
    elif 'sma/smk' in desc:
        counter += 1
    elif 'sarjana' in desc:
        counter+=1
    elif 'fresh graduate' in desc:
        counter += 1
    elif 'smk' in desc:
        counter += 1
    elif 'sma/k' in desc:
        counter+=1
    # else:
    #     print(desc)
counter

925

In [35]:
list_pendidikan = ['sma/smk', 'sma/k', 'smk', 'diploma', 'd3', 'sarjana','s1',
                   'bachelor', 'diploma', 'master degree', 'masters degree', ' fresh graduate']
pendidikan_mapper = {'sma/smk':'sma/smk',
                     'sma/k':'sma/smk', 
                     'smk':'smk',
                     'diploma':'d3',
                     'd3':'d3',
                     'sarjana':'s1',
                     'bachelor':'s1',
                     'master degree':'s2',
                     'masters degree':'s2',
                     'fresh graduate':'d3/s1'}
                     

extract_pendidikan = list()
td = 0
for job_desc in list(job_description.values):
    # replace token [sep]
    job_desc = job_desc.replace('[sep]', ' ')
    job_desc = job_desc.replace(':', ' ')

    # split dengan space
    split_desc = job_desc.split()

    item_match = list()
    for item in split_desc:
        if any(x in item for x in list_pendidikan):
            # cleaning punctuation
            item = re.sub(r"[(,.;@#?!&$)/]+\ *", " ", item).strip()
            
            # split from 's
            item = item.split("'")[0]
            item = item.split("’")[0]

            # split
            for it in item.split():
                item_match.append(it)
    
    # mapping pendidikan
    item_match = pd.Series(item_match).map(pendidikan_mapper).dropna().drop_duplicates().tolist()

    if len(item_match) >=1:
        extract_pendidikan.append(item_match)
    else:
        extract_pendidikan.append(['tidak dicantumkan'])
        td+=1
td

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:40: DeprecationWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.


1064

In [36]:
# join

for idx, pend in enumerate(extract_pendidikan):
    join = ', '.join(pend).strip()
    extract_pendidikan[idx] = join

In [37]:
glints['pendidikan'] = extract_pendidikan

In [38]:
glints.head()

platform         job_position  ...        provinsi         pendidikan
0   glints       .Net Developer  ...  Kepulauan Riau                 d3
1   glints  Fullstack Developer  ...  Kepulauan Riau                 d3
2   glints   Frontend Developer  ...     DKI Jakarta  tidak dicantumkan
3   glints    Backend Developer  ...     DKI Jakarta  tidak dicantumkan
4   glints   Frontend Developer  ...  Kepulauan Riau  tidak dicantumkan

[5 rows x 12 columns]

## skills_to_have

In [39]:
skills_to_have = glints['skills_to_have']
skills_to_have

0       Must Have Skills [SEP] .NET [SEP] English [SEP...
1       Must Have Skills [SEP] ReactJS [SEP] HTML [SEP...
2       Must Have Skills [SEP] Teamwork [SEP] Microsof...
3       Must Have Skills [SEP] Agile Scrum [SEP] Engli...
4       Must Have Skills [SEP] Cms Wordpress [SEP] Wor...
                              ...                        
1744    Must Have Skills [SEP] Graphic Design [SEP] At...
1745    Must Have Skills [SEP] Market Reserach [SEP] B...
1746    Must Have Skills [SEP] Node.js [SEP] Nodejs Ex...
1747    Must Have Skills [SEP] Digital Marketing [SEP]...
1748    Must Have Skills [SEP] English [SEP] Project M...
Name: skills_to_have, Length: 1749, dtype: object

In [40]:
# hapus yang bukan skill
result = list()
delete = ['Must Have Skills', 'Nice To Have Skills']
for skill in skills_to_have:
    split = [data.strip() for data in skill.split('[SEP]')]
    
    temp = list()
    for s in split:
        if s in delete:
            continue
        temp.append(s)
    join = ", ".join(temp).strip()
    result.append(join)

In [41]:
result[0]

'.NET, English, Ado.net, Windows Azure, Saas, COM+, ASP.NET, VB.NET, HTML, XML, XPath'

In [42]:
glints['skills_to_have'] = result

In [43]:
glints.head()

platform         job_position  ...        provinsi         pendidikan
0   glints       .Net Developer  ...  Kepulauan Riau                 d3
1   glints  Fullstack Developer  ...  Kepulauan Riau                 d3
2   glints   Frontend Developer  ...     DKI Jakarta  tidak dicantumkan
3   glints    Backend Developer  ...     DKI Jakarta  tidak dicantumkan
4   glints   Frontend Developer  ...  Kepulauan Riau  tidak dicantumkan

[5 rows x 12 columns]

# Export csv

In [44]:
glints.head()

platform         job_position  ...        provinsi         pendidikan
0   glints       .Net Developer  ...  Kepulauan Riau                 d3
1   glints  Fullstack Developer  ...  Kepulauan Riau                 d3
2   glints   Frontend Developer  ...     DKI Jakarta  tidak dicantumkan
3   glints    Backend Developer  ...     DKI Jakarta  tidak dicantumkan
4   glints   Frontend Developer  ...  Kepulauan Riau  tidak dicantumkan

[5 rows x 12 columns]

In [45]:
# drop job_desc
glints = glints.drop(['job_desc'], axis=1)
glints.head()

platform         job_position  ...        provinsi         pendidikan
0   glints       .Net Developer  ...  Kepulauan Riau                 d3
1   glints  Fullstack Developer  ...  Kepulauan Riau                 d3
2   glints   Frontend Developer  ...     DKI Jakarta  tidak dicantumkan
3   glints    Backend Developer  ...     DKI Jakarta  tidak dicantumkan
4   glints   Frontend Developer  ...  Kepulauan Riau  tidak dicantumkan

[5 rows x 11 columns]

In [46]:
# tambah id_loker
id_loker = [i+1 for i in range(len(glints))]
glints['id_loker'] = id_loker

glints.head()
# glints.to_csv("glints_15oct.csv", index=False)

platform         job_position  ...         pendidikan id_loker
0   glints       .Net Developer  ...                 d3        1
1   glints  Fullstack Developer  ...                 d3        2
2   glints   Frontend Developer  ...  tidak dicantumkan        3
3   glints    Backend Developer  ...  tidak dicantumkan        4
4   glints   Frontend Developer  ...  tidak dicantumkan        5

[5 rows x 12 columns]

In [47]:
# rearrange cols
cols = ['id_loker',
        'pulau',
        'provinsi',
        'kota',
        'job_position',
        'skills_to_have',
        'pendidikan',
        'company_name',
        'job_category',
        'job_salary',
        'job_type',
        'platform']
glints = glints[cols]

glints.head()

id_loker     pulau  ...    job_type platform
0         1  Sumatera  ...   FULL_TIME   glints
1         2  Sumatera  ...   FULL_TIME   glints
2         3      Jawa  ...  INTERNSHIP   glints
3         4      Jawa  ...   FULL_TIME   glints
4         5  Sumatera  ...   FULL_TIME   glints

[5 rows x 12 columns]

## Export data_loker_glints

In [48]:
# export data_loker
glints.to_csv("data_loker_glints_16oct.csv", index=False)

## Export data_pendidikan_glints

In [49]:
# reformat data pendidikan menjadi csv baru
df_pendidikan = pd.DataFrame(columns=['id_loker', 'provinsi', 'pendidikan'])

for idx, row in glints.iterrows():

    split_pendidikan = [data.strip().upper() for data in row['pendidikan'].split(',')]
    
    for pend in split_pendidikan:
        df_pendidikan = df_pendidikan.append({'id_loker': row['id_loker'],
                                              'provinsi': row['provinsi'],
                                              'pendidikan': pend}, ignore_index=True)

In [50]:
df_pendidikan.to_csv('data_pendidikan_glints_16oct.csv', index=False)

## Export data_skill_glints

In [51]:
# reformat data skill menjadi csv baru
df_skill = pd.DataFrame(columns=['id_loker', 'provinsi', 'skills'])

for idx, row in glints.iterrows():

    split_skill = [data.strip() for data in row['skills_to_have'].split(',')]
    
    for skill in split_skill:
        df_skill = df_skill.append({'id_loker': row['id_loker'],
                                    'provinsi': row['provinsi'],
                                    'skills': skill}, ignore_index=True)

In [52]:
df_skill.head()

id_loker        provinsi         skills
0        1  Kepulauan Riau           .NET
1        1  Kepulauan Riau        English
2        1  Kepulauan Riau        Ado.net
3        1  Kepulauan Riau  Windows Azure
4        1  Kepulauan Riau           Saas

In [53]:
df_skill.shape

(11641, 3)

In [54]:
df_skill.to_csv("data_skill_glints_16oct.csv", index=False)